File: 1-Part/6-chapter/balanceParentheses.js
function getBalancedParens(pairs, openRem, closeRem, current, indent) {
    if (openRem === undefined) {
        openRem = pairs
    }
    if (closeRem === undefined) {
        closeRem = pairs
    }
    if (current === undefined) {
        current = ''
    }
    if (indent === undefined) {
        indent = 0
    }
    console.log(`${'.'.repeat(indent)} Start of pairs=${pairs}, openRem=${openRem}, closeRem=${closeRem}, current=${current}`)

    // Base case
    if (openRem === 0 && closeRem === 0) {
        console.log(`${'.'.repeat(indent)} 1-st base case. Returning ${[current]}`)
        return [current]
    }
//     Recursion case
    let results = []
    if (openRem > 0) {
        console.log(`${'.'.repeat(indent)} Adding open parenthesis`)
        Array.prototype.push.apply(results, getBalancedParens(pairs, openRem - 1, closeRem, current + '(', indent + 1))
    }
    if (closeRem > openRem) {
        console.log(`${'.'.repeat(indent)} Adding close parenthesis`)
        results = results.concat(getBalancedParens(pairs, openRem, closeRem - 1, current + ')', indent + 1))
    }
//     Base case
    console.log(`${'.'.repeat(indent)} 2nd base case. Returning ${results}`)
    return results
}

console.log('All combinations of 2 balanced parentheses:')
console.log('Results:', getBalancedParens(2))

File: 1-Part/6-chapter/balanceParentheses.py
def getBalanceParens(pairs, openRem=None, closeRem=None, current='', indent=0):
    if openRem is None:
        openRem = pairs
    if closeRem is None:
        closeRem = pairs

    print('.' * indent, end='')
    print('Start of pairs=' + str(pairs) + ',openRem = ' + str(openRem) + ', closeRem= ' + str(
        closeRem) + ', current="' + current + '" ')
    if openRem == 0 and closeRem == 0:
        print('.' * indent, end='')
        print('1st base case. Returning' + str([current]))
        return [current]

    results = []
    if openRem > 0:
        print('.' * indent, end='')
        print('Adding open parenthesis')
        results.extend(getBalanceParens(pairs, openRem - 1, closeRem, current + '(', indent + 1))
    if closeRem > openRem:
        print('.' * indent, end='')
        print('Adding close parenthesis')
        results.extend(getBalanceParens(pairs, openRem, closeRem - 1, current + ')', indent + 1))

    print('.' * indent, end='')
    print('2st base case. Returning' + str(results))
    return results


print('All combinations of 2 balanced parentheses:')
print('Results:', getBalanceParens(2))
